### PR TITLE
note/sysview: add Segger System View support

### PR DESCRIPTION
**Pull request by anchao**
## Summary

sim/sysview: add build config of Segger System View
segger/RTT: hotfix for RTT compile warning
drivers/segger: download the RTT/SystemView from github
note/sysview: add Segger System View support
tools/nxstyle: add Segger System View prototype name into white list
sched/critmonitor: remove the assertion if counter == 0
timers/critmon: mark the weak reference for up_critmon_*

Reference:
https://www.segger.com/products/development-tools/systemview/
https://github.com/SEGGERMicro/SystemView
https://github.com/SEGGERMicro/RTT

## Impact

N/A, new feature

## Testing

Test on Cortex-m55:

https://www.segger.com/downloads/systemview/

Screenshot from 2022-01-25 00-02-19

### CODE REVIEW

**Comment by acassis** [comment]
hi @anchao did you test it on more common boards like stm32f4discovery? Unfortunately Segger doesn't support ESP32 yet!

**Comment by anchao** [comment]
> hi @anchao did you test it on more common boards like stm32f4discovery? Unfortunately Segger doesn't support ESP32 yet!
@acassis ,
I only tested the arm architecture, the porting layer does not use architecture-related implementations from RTT, I think xtensa should also be works fine without any changes

**Comment by acassis** [comment]
Thank you @anchao I just saw you closed this PR, did you find some issue?

**Comment by anchao** [comment]
> Thank you @anchao I just saw you closed this PR, did you find some issue?

@acassis  

This commit relies on tweaks implemented by up_perf* related apis，@xiaoxiang781216 has completed the improvement, I will reopen this PR asap

File: drivers/segger/Make.defs [hunk 1]
```diff
@@ -38,4 +51,21 @@ ifeq ($(CONFIG_SEGGER_SYSVIEW),y)
   CFLAGS += ${shell $(INCDIR) "$(CC)" $(TOPDIR)$(DELIM)sched}
   CFLAGS += ${shell $(INCDIR) "$(CC)" segger$(DELIM)SystemView$(DELIM)SEGGER}
   CFLAGS += ${shell $(INCDIR) "$(CC)" segger$(DELIM)SystemView$(DELIM)SYSVIEW}
+
+  SYSVIEW_VERSION ?= 3.30
```
Comment (xiaoxiang781216) [inline drivers/segger/Make.defs:56]: remove  the space

File: drivers/segger/Make.defs [hunk 2]
```diff
@@ -29,6 +32,16 @@ ifeq ($(CONFIG_SEGGER_RTT),y)
   ifeq ($(CONFIG_ARCH_ARMV7M),y)
     ASRCS += segger/RTT/RTT/SEGGER_RTT_ASM_ARMv7M.S
   endif
+
+  RTT_VERSION     ?= 7.54
```
Comment (xiaoxiang781216) [inline drivers/segger/Make.defs:36]: remove space

Comment (anchao) [reply drivers/segger/Make.defs]: Done

**Review by xiaoxiang781216 — APPROVED** [review]
Ok, LGTM.

**Review by a-lunev — COMMENTED** [review]
.

File: drivers/segger/config/Global.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
+/****************************************************************************
+ * drivers/segger/config/Global.h
```
Comment (a-lunev) [inline drivers/segger/config/Global.h:2]: I think it'd better to lower case the file name (Global.h -> global.h) as almost all NuttX file names are lower case.

Comment (anchao) [reply drivers/segger/config/Global.h:2]: drivers/segger/config dir is not use for nuttx, this is overwrite include path for system view:

https://github.com/SEGGERMicro/SystemView/blob/master/SEGGER/SEGGER.h#L59
https://github.com/SEGGERMicro/SystemView/blob/master/Config/Global.h

File: tools/nxstyle.c [hunk 4]
```diff
@@ -307,6 +307,17 @@ static const char *g_white_list[] =
   "__asan_loadN_noabort",
   "__asan_storeN_noabort",
 
+  /* Ref:
+   * drivers/segger/note_sysview.c
```
[the remark this answers is not in the document]
Comment (anchao) [reply tools/nxstyle.c]: Done

**Review by a-lunev — APPROVED** [review]

File: drivers/segger/config/SEGGER_SYSVIEW_Conf.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+#define __DRIVERS_SEGGER_CONFIG_SEGGER_SYSVIEW_CONF_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+extern unsigned int sysview_get_interrupt_id(void);
+extern unsigned int sysview_get_cycle_counter(void);
```
Comment (pkarashchenko) [inline drivers/segger/config/SEGGER_SYSVIEW_Conf.h:31]: Maybe should be moved to a separate section?

File: drivers/segger/config/SEGGER_SYSVIEW_Conf.h [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+#define SEGGER_SYSVIEW_RTT_BUFFER_SIZE     CONFIG_SEGGER_SYSVIEW_RTT_BUFFER_SIZE
+
+/* Largest cache line size (in bytes) in the target system. */
+
+#define SEGGER_SYSVIEW_CPU_CACHE_LINE_SIZE CONFIG_SEGGER_RTT_CPU_CACHE_LINE_SIZE
+
+/* Enable calling SEGGER_SYSVIEW_Start() after initialization. */
+
+#ifdef CONFIG_SEGGER_SYSVIEW_START_ON_INIT
+  #define SEGGER_SYSVIEW_START_ON_INIT     1
```
Comment (pkarashchenko) [inline drivers/segger/config/SEGGER_SYSVIEW_Conf.h:60]:
```suggestion
#  define SEGGER_SYSVIEW_START_ON_INIT     1
```

File: drivers/segger/note_sysview.c [hunk 7]
```diff
@@ -0,0 +1,553 @@
...
+
+#include "sched/sched.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#ifndef CONFIG_SMP_NCPUS
+#define CONFIG_SMP_NCPUS 1
+#endif
```
Comment (pkarashchenko) [inline drivers/segger/note_sysview.c:44]:
```suggestion
```
Already available from `sys/types.h`

File: drivers/segger/note_sysview.c [hunk 8]
```diff
@@ -0,0 +1,553 @@
...
+    }
+}
+
+/****************************************************************************
+ * Name: sysview_send_description
+ ****************************************************************************/
+
+static void sysview_send_description(void)
+{
+  SEGGER_SYSVIEW_SendSysDesc("N="SEGGER_SYSVIEW_APP_NAME"");
```
Comment (pkarashchenko) [inline drivers/segger/note_sysview.c:138]:
```suggestion
  SEGGER_SYSVIEW_SendSysDesc("N="SEGGER_SYSVIEW_APP_NAME);
```
??  Not sure why `""` is needed

File: drivers/segger/note_sysview.c [hunk 9]
```diff
@@ -0,0 +1,553 @@
...
+  SEGGER_SYSVIEW_SendSysDesc("D="CONFIG_LIBC_HOSTNAME"");
```
Comment (pkarashchenko) [inline drivers/segger/note_sysview.c:139]:
```suggestion
  SEGGER_SYSVIEW_SendSysDesc("D="CONFIG_LIBC_HOSTNAME);
```
??  Not sure why `""` is needed

File: drivers/segger/note_sysview.c [hunk 10]
```diff
@@ -0,0 +1,553 @@
...
+  if (enable)
+    {
+      if ((g_sysview.mode.flag & NOTE_FILTER_MODE_FLAG_ENABLE) == 0)
+        {
+          g_sysview.mode.flag |= NOTE_FILTER_MODE_FLAG_ENABLE;
+        }
+    }
+  else
+    {
+      if (g_sysview.mode.flag & NOTE_FILTER_MODE_FLAG_ENABLE)
```
Comment (pkarashchenko) [inline drivers/segger/note_sysview.c:173]:
```suggestion
      if ((g_sysview.mode.flag & NOTE_FILTER_MODE_FLAG_ENABLE) != 0)
```

File: drivers/segger/note_sysview.c [hunk 11]
```diff
@@ -0,0 +1,553 @@
...
+  if (!sysview_isenabled())
+    {
+      return false;
+    }
+
+  /* If the IRQ trace is disabled or the IRQ number is masked, disable
+   * subsequent syscall traces until leaving the interrupt handler
+   */
+
+  if (!(g_sysview.mode.flag & NOTE_FILTER_MODE_FLAG_IRQ) ||
```
Comment (pkarashchenko) [inline drivers/segger/note_sysview.c:226]:
```suggestion
  if ((g_sysview.mode.flag & NOTE_FILTER_MODE_FLAG_IRQ) == 0 ||
```

File: drivers/segger/note_sysview.c [hunk 12]
```diff
@@ -0,0 +1,553 @@
...
+  struct sysview_cpu_s *scpu;
+
+  if (!sysview_isenabled())
+    {
+      return;
+    }
+
+  scpu = &g_sysview.scpu[up_cpu_index()];
+
+  if (scpu->in_isrcb == false)
```
Comment (pkarashchenko) [inline drivers/segger/note_sysview.c:302]:
```suggestion
  if (!scpu->in_isrcb)
```

File: drivers/segger/note_sysview.c [hunk 13]
```diff
@@ -0,0 +1,553 @@
...
+  if (!sysview_isenabled_irq(irq, enter))
+    {
+      return;
+    }
+
+  scpu = &g_sysview.scpu[up_cpu_index()];
+
+  if (enter)
+    {
+      if (g_sysview.mode.flag & NOTE_FILTER_MODE_FLAG_ENABLE)
```
Comment (pkarashchenko) [inline drivers/segger/note_sysview.c:336]:
```suggestion
      if ((g_sysview.mode.flag & NOTE_FILTER_MODE_FLAG_ENABLE) != 0)
```

File: drivers/segger/note_sysview.c [hunk 14]
```diff
@@ -0,0 +1,553 @@
...
+
+  SEGGER_SYSVIEW_Init(freq, freq, &g_sysview_trace_api,
+                      sysview_send_description);
+
+#if CONFIG_SEGGER_SYSVIEW_RAM_BASE != 0
+  SEGGER_SYSVIEW_SetRAMBase(CONFIG_SEGGER_SYSVIEW_RAM_BASE);
+#endif
+
+#ifdef CONFIG_SEGGER_SYSVIEW_START_ON_INIT
+  if (g_sysview.mode.flag & NOTE_FILTER_MODE_FLAG_ENABLE)
```
Comment (pkarashchenko) [inline drivers/segger/note_sysview.c:438]:
```suggestion
  if ((g_sysview.mode.flag & NOTE_FILTER_MODE_FLAG_ENABLE) != 0)
```

File: drivers/segger/config/SEGGER_SYSVIEW_Conf.h (hunk 6 above)
Comment (anchao) [reply drivers/segger/config/SEGGER_SYSVIEW_Conf.h]: done

File: drivers/segger/note_sysview.c (hunk 7 above)
Comment (anchao) [reply drivers/segger/note_sysview.c]: done

File: drivers/segger/note_sysview.c (hunk 9 above)
Comment (anchao) [reply drivers/segger/note_sysview.c:123]: done

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@pkarashchenko do you have more comment?

**Review by pkarashchenko — APPROVED** [review]
There is one minor comment that is still not addressed, but nothing that blocks from merging.

File: drivers/segger/note_sysview.c (hunk 11 above)
Comment (anchao) [reply drivers/segger/note_sysview.c:189]: done

**Comment by anchao** [comment]
> There is one minor comment that is still not addressed, but nothing that blocks from merging.

resolved

**Review by pkarashchenko — APPROVED** [review]
LGTM!

**Pull request merged**